lib/src/signed_video_openssl.c

```diff
@@ -319,51 +320,80 @@ openssl_hash_data(void *handle, const uint8_t *data, size_t data_size, uint8_t *
 
 /* Initializes EVP_MD_CTX in |handle| with |hash_algo.type|. */
 svrc_t
-openssl_init_hash(void *handle)
+openssl_init_hash(void *handle, bool fallback_to_gop_level)
 {
   if (!handle) return SV_INVALID_PARAMETER;
   openssl_crypto_t *self = (openssl_crypto_t *)handle;
   int ret = 0;
 
-  if (self->ctx) {
-    // Message digest type already set in context. Initialize the hashing function.
-    ret = EVP_DigestInit_ex(self->ctx, NULL, NULL);
+  if (fallback_to_gop_level) {
+    if (self->f_ctx) {
+      // Message digest type already set in context. Initialize the hashing function.
+      ret = EVP_DigestInit_ex(self->f_ctx, NULL, NULL);
+    } else {
+      if (!self->hash_algo.type) return SV_INVALID_PARAMETER;
+      // Create a new context and set message digest type.
+      self->f_ctx = EVP_MD_CTX_new();
+      if (!self->f_ctx) return SV_EXTERNAL_ERROR;
+      // Set a message digest type and initialize the hashing function.
+      ret = EVP_DigestInit_ex(self->f_ctx, self->hash_algo.type, NULL);
+    }
   } else {
-    if (!self->hash_algo.type) return SV_INVALID_PARAMETER;
-    // Create a new context and set message digest type.
-    self->ctx = EVP_MD_CTX_new();
-    if (!self->ctx) return SV_EXTERNAL_ERROR;
-    // Set a message digest type and initialize the hashing function.
-    ret = EVP_DigestInit_ex(self->ctx, self->hash_algo.type, NULL);
+    if (self->ctx) {
+      // Message digest type already set in context. Initialize the hashing function.
+      ret = EVP_DigestInit_ex(self->ctx, NULL, NULL);
+    } else {
+      if (!self->hash_algo.type) return SV_INVALID_PARAMETER;
+      // Create a new context and set message digest type.
+      self->ctx = EVP_MD_CTX_new();
+      if (!self->ctx) return SV_EXTERNAL_ERROR;
+      // Set a message digest type and initialize the hashing function.
+      ret = EVP_DigestInit_ex(self->ctx, self->hash_algo.type, NULL);
+    }
```

Comment:
Duplicating code this way makes the code less readable and becomes very vulnerable to introducing bugs, since you need to make the same change in several places. Good coding practice is to select the context to us first and then proceed with the operation, that is,
```
EVP_MD_CTX **ctx = primary ? &self->primary_ctx : &self->secondary_ctx;
if (*ctx) {
  // Message digest type already set in context. Initialize the hashing function.
  ret = EVP_DigestInit_ex(*ctx, NULL, NULL);
} else {
  if (!self->hash_algo.type) return SV_INVALID_PARAMETER;
  // Create a new context and set message digest type.
  *ctx = EVP_MD_CTX_new();
  if (!(*ctx)) return SV_EXTERNAL_ERROR;
  // Set a message digest type and initialize the hashing function.
  ret = EVP_DigestInit_ex(*ctx, self->hash_algo.type, NULL);
}
```